src/Microsoft.Health.Fhir.Core/Features/Operations/DataConvert/ContainerRegistryTemplateProvider.cs

```diff
@@ -0,0 +1,105 @@
+using Microsoft.Extensions.Logging;
+using Microsoft.Health.Fhir.Core.Features.Operations.DataConvert.Models;
+using Microsoft.Health.Fhir.TemplateManagement;
+using Microsoft.Health.Fhir.TemplateManagement.Exceptions;
+using Microsoft.Health.Fhir.TemplateManagement.Models;
+
+namespace Microsoft.Health.Fhir.Core.Features.Operations.DataConvert
+{
+    public class ContainerRegistryTemplateProvider : IDataConvertTemplateProvider
+    {
+        private readonly IContainerRegistryTokenProvider _containerRegistryTokenProvider;
+        private readonly ITemplateSetProviderFactory _templateSetProviderFactory;
+        private readonly ILogger<ContainerRegistryTemplateProvider> _logger;
+
+        public ContainerRegistryTemplateProvider(
+            IContainerRegistryTokenProvider containerRegistryTokenProvider,
+            ITemplateSetProviderFactory templateSetProviderFactory,
+            ILogger<ContainerRegistryTemplateProvider> logger)
+        {
+            EnsureArg.IsNotNull(containerRegistryTokenProvider, nameof(containerRegistryTokenProvider));
+            EnsureArg.IsNotNull(templateSetProviderFactory, nameof(templateSetProviderFactory));
+            EnsureArg.IsNotNull(logger, nameof(logger));
+
+            _containerRegistryTokenProvider = containerRegistryTokenProvider;
+            _templateSetProviderFactory = templateSetProviderFactory;
+            _logger = logger;
+        }
+
+        /// <summary>
+        /// Fetch template collection from container registry or built-in archive
+        /// </summary>
+        /// <param name="templateCollectionReference">Should be in "<registryServer>/<imageName>:<imageTag>" or "<registryServer>/<imageName>@<imageDigest>"  format</param>
+        /// <param name="cancellationToken">Cancellation token to cancel the fetch operation.</param>
+        /// <returns>Template collection.</returns>
+        public async Task<List<Dictionary<string, Template>>> GetTemplateCollectionAsync(string templateCollectionReference, CancellationToken cancellationToken)
+        {
+            // We have embedded a default template set in the templatemanagement package.
+            // If the template set is the default reference, we don't need to retrieve token.
+            var registryServer = ExtractRegistryServer(templateCollectionReference);
+            var accessToken = string.Empty;
+            if (!IsDefaultTemplateReference(templateCollectionReference))
+            {
+                _logger.LogInformation("Using a custom template set for data conversion.");
+                accessToken = await _containerRegistryTokenProvider.GetTokenAsync(registryServer, cancellationToken);
+            }
+            else
+            {
+                _logger.LogInformation("Using the default template set for data conversion.");
+            }
+
+            try
+            {
+                var provider = _templateSetProviderFactory.CreateTemplateSetProvider(templateCollectionReference, accessToken);
```

Comment:
>_templateSetProviderFactory.CreateTemplateSetProvider(templateCollectionReference [](start = 31, length = 81)

super nit:
It's a bit strange to see _template*Set*Provider to work with template*Collection*.
I think it would be nice to align them

Reply:
Sure. Will change to templateCollection